feat: Add Xen support

andreistan26 opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [ ] Tested your changes against relevant architectures and platforms;
  - [ ] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

GitHub-Closes: #990

andreistan26 commented on internal/logtail/logtail.go:
```diff
@@ -64,7 +64,11 @@ func NewLogTail(ctx context.Context, logFile string) (chan string, chan error, e
 				switch event.Op {
 				case fsnotify.Write:
-					peekAndRead(f, reader, &logs, &errs)
+					for {
```
We can get multi-line writes with one event triggered. 
IMO something better would be for `peekAndRead` to look ahead for the last `newline` so we end up with a single write.

andreistan26 commented on internal/logtail/logtail.go:
```diff
@@ -64,7 +64,11 @@ func NewLogTail(ctx context.Context, logFile string) (chan string, chan error, e
 				switch event.Op {
 				case fsnotify.Write:
-					peekAndRead(f, reader, &logs, &errs)
+					for {
+						if peekAndRead(f, reader, &logs, &errs) {
+							break
+						}
+					}
```
We can get multi-line writes with one event triggered. 
IMO something better would be for `peekAndRead` to look ahead for the last `newline` so we end up with a single write.

andreistan26 committed:
feat(machine/xen): Introduce Xen support

Signed-off-by: Andrei Stan <andreistan2003@gmail.com>

andreistan26 committed:
fix(logtail): Output multiline text on single inotify event

Signed-off-by: Andrei Stan <andreistan2003@gmail.com>

craciunoiuc commented on buildenvs/xen.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022, Unikraft GmbH and The KraftKit Authors.
```
```suggestion
# Copyright (c) 2024, Unikraft GmbH and The KraftKit Authors.
```

Same for others

craciunoiuc commented on buildenvs/xen.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+RUN set -xe; \
+    apt-get update; \
+    apt-get install -y \
+        cmake \
+        libssh2-1-dev \
+        binutils \
+        bison \
+        build-essential \
+        flex \
+        gcc \
+        git \
+        libncurses5-dev \
+        libssl-dev \
+        make \
+        python3 \
+        python3-pip \
+        python3-setuptools \
+        python3-wheel \
+        ninja-build \
+        perl \
+        pkg-config \
+        libglib2.0-dev \
+        libpixman-1-dev \
+        iasl \
+        libyajl-dev \
+        uuid-dev \
+        libslirp-dev; \
```
Order alphabetically please :pray:

craciunoiuc commented on buildenvs/xen.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+# Licensed under the BSD-3-Clause License (the "License").
+# You may not use this file except in compliance with the License.
+
+FROM kraftkit.sh/base-golang:latest
```
shouldnt this be based on `base`? just asking

andreistan26 replied:
`base-golang` is the same as `base` with Go installed.

craciunoiuc commented on machine/platform/register_linux.go:
```diff
@@ -43,6 +44,30 @@ var firecrackerV1alpha1Driver = func(ctx context.Context, opts ...any) (machinev
 	)
 }
 
+var xenV1alpha1Driver = func(ctx context.Context, opts ...any) (machinev1alpha1.MachineService, error) {
+	service, err := xen.NewMachineV1alpha1Service(ctx)
+	if err != nil {
+		return nil, err
+	}
```
research if xen works fine (compiles) also on `openbsd`/`freebsd` I think you can find dockerfiles for them.

You can then add equivalent `register_openbsd.go` and `register_freebsd.go` files.

Also I guess `darwin` is not the case currently for `xen`

andreistan26 replied:
For `bsd` builds i would need a vm in order to test it. I think it would be better to leave that after the upstream of the `linux` only feature.

craciunoiuc replied:
sure thing

craciunoiuc commented on machine/xen/stub.go:
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2022, Unikraft GmbH and The KraftKit Authors.
```
```suggestion
// Copyright (c) 2024, Unikraft GmbH and The KraftKit Authors.
```

This for all files

craciunoiuc commented on machine/xen/xenstore_client.go:
```diff
@@ -0,0 +1,194 @@
+type XenstoreOperation uint32
+
+const (
+	WatchOp    XenstoreOperation = 4
+	UnwatchOp  XenstoreOperation = 5
+	WatchEvent XenstoreOperation = 15
+	Error      XenstoreOperation = 16
```
Are these bit flags? if yes you could use `0b...` to specify them

andreistan26 replied:
These are request/response codes for `xenstore`, not flags.

craciunoiuc commented on machine/xen/xenstore_client.go:
```diff
@@ -0,0 +1,194 @@
+)
+
+func XenstoreSocketPath() string {
+	xenstorepath := "/var/run/xenstored"
```
Make this path also a default in the consts

craciunoiuc commented on machine/xen/xenstore_client.go:
```diff
@@ -0,0 +1,194 @@
+	}, nil
+}
+
+func (packet *XsPacket) pack() []byte {
+	data := make([]byte, 0)
+	data = binary.LittleEndian.AppendUint32(data, uint32(packet.Header.Op))
+	data = binary.LittleEndian.AppendUint32(data, packet.Header.ReqID)
+	data = binary.LittleEndian.AppendUint32(data, packet.Header.TxID)
+	data = binary.LittleEndian.AppendUint32(data, packet.Header.Lenght)
+
+	data = append(data, packet.Data...)
+
+	return data
+}
+
+func unpack(data []byte) XsPacket {
```
I think you'll need a nice ascii comment to explain these 2 functions for future maintenance

craciunoiuc commented on machine/xen/xenstore_client.go:
```diff
@@ -0,0 +1,194 @@
+	}
+	packet := XsPacket{
+		Header: header,
+		Data:   data[16 : 16+header.Lenght],
```
`Length`

craciunoiuc commented on machine/xen/xenstore_client.go:
```diff
@@ -0,0 +1,194 @@
+
+func (w *baseWatcher) xsWatchRequest() error {
+
+	data := append(append(append([]byte(w.xsPath), '\x00'), []byte(w.token)...), '\x00')
```
This looks a bit nasty, can you prettify it a bit?

craciunoiuc commented on machine/xen/xenstore_client.go:
```diff
@@ -0,0 +1,194 @@
+
+func SplitData(packet XsPacket) []string {
+	splitPayload := []string{}
+	for _, byteSl := range bytes.Split(packet.Data, []byte("\x00")) {
```
what's with the `"\x00"` bytes everywhere, maybe you can move them to a constant or find the `NUL` character somewhere in some golang library

andreistan26 replied:
The `\x00` are required for the payloads of the messages  to `xenstore`. Converting a string to a byte slice in Go does not add the null byte at the end, it has to be done explicitly.

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+		return machine, fmt.Errorf("cannot create xen instace without a kernel")
+	}
+
+	if !slices.Contains([]string{"arm64", "x86_64", "amd64", "arm"}, machine.Spec.Architecture) {
```
I think you can get this list from a function in kraftkit, try to look for it I can't remember

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+	}
+
+	if machine.Spec.Resources.Requests.Memory().Value() == 0 {
+		quantity, err := resource.ParseQuantity(fmt.Sprintf("%d", XenMemoryDefault) + "Mi")
```
```suggestion
		quantity, err := resource.ParseQuantity(fmt.Sprintf("%d Mi", XenMemoryDefault))
```

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+	}
+
+	cfg.BInfo.MaxVcpus = int(machine.Spec.Resources.Requests.Cpu().Value())
+	cfg.BInfo.MaxMemkb = uint64(machine.Spec.Resources.Requests.Memory().Value() / XenMemoryScale)
```
A corner case, but, if you want to start a machine with `<1 Ki` this division will make it have 0 

Can you do a check above to see, if memory was specified, that it's at least `1KiB` :sweat_smile:

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+	cfg.CInfo.Name = string(machine.ObjectMeta.Name)
+	cfg.CInfo.Type = xenlight.DomainTypePv
+
+	if len(machine.Status.InitrdPath) > 0 {
```
```suggestion
	if machine.Status.InitrdPath != "" {
```

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+	// TODO(andreistan26): Check if installed xen supports 9pfs
+	for i, vol := range machine.Spec.Volumes {
+		switch vol.Spec.Driver {
+		case "9pfs":
```
not sure if value is from an `enum`. If not, then I would match also `9p` for now

andreistan26 replied:
`vol.Spec.Driver` is a string, i've found no enum for `volume` types

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+	}
+	defer xenCtx.Close()
+
+	if err := xenCtx.DomainUnpause(domId); err != nil {
```
Why does `Pause` do `Unpausing` is it a typo?

andreistan26 replied:
Nice catch, forgot to test the `pause` command. That should be `DomainPause`.

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+	return machine, nil
+}
+func (service *machineV1alpha1Service) Update(ctx context.Context, machine *machinev1alpha1.Machine) (*machinev1alpha1.Machine, error) {
+	panic("not implemented: kraftkit.sh/machine/xen.machineV1alpha1Service.Update")
```
Is this how it's done in the other `v1alpha1.go` files?

andreistan26 replied:
Yes

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+	panic("not implemented: kraftkit.sh/machine/xen.machineV1alpha1Service.Update")
+}
+func (service *machineV1alpha1Service) Delete(ctx context.Context, machine *machinev1alpha1.Machine) (*machinev1alpha1.Machine, error) {
+	err := os.RemoveAll(machine.Status.StateDir)
```
Not sure but maybe `Delete` should also check if the machine is stopped first

andreistan26 replied:
That is done by the caller, within `remove` or when `--rm` is passed. Neither `qemu` nor `fc` check that. So I think it should be left as it is.

craciunoiuc commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,536 @@
+	return logtail.NewLogTail(ctx, machine.Status.LogFile)
+}
+
+func getXenState(domInfo *xenlight.Dominfo) (int, machinev1alpha1.MachineState) {
```
Add a comment explaining `-1/0/1`, or even better turn the `int` into an enum `exitStatus` or something like that

`type exitStatus int`

andreistan26 replied:
To be honest i am not sure about the spec for the exit codes, these are the same as for `qemu`.

craciunoiuc replied:
sounds good then

craciunoiuc reviewed (changes requested):
Some style comments from my side, I did not try it yet

andreistan26 committed:
build(xen): Add container for Xen enabled builds and update goreleaser

Signed-off-by: Andrei Stan <andreistan2003@gmail.com>

andreistan26 committed:
fix(logs): Tail logs from already exited unikernel

In the case of Xen, on fast applications, the logs
would have been called when the machine was already
in the exit state.

Signed-off-by: Andrei Stan <andreistan2003@gmail.com>

craciunoiuc commented on buildenvs/xen.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+        python3-wheel \
+        uuid-dev; \
+    pip3 install python-config --break-system-packages; \
+    git clone -b stable-4.18 https://xenbits.xen.org/git-http/xen.git /xen; \
```
can you turn this version into an `ARG`?

```suggestion
    git clone -b stable-${XEN_VERSION} https://xenbits.xen.org/git-http/xen.git /xen; \
```

craciunoiuc commented on internal/cli/kraft/logs/logs.go:
```diff
@@ -167,7 +167,9 @@ func (opts *LogOptions) Run(ctx context.Context, args []string) error {
-		if opts.Follow && machine.Status.State == machineapi.MachineStateRunning {
+
+		// Sometimes the kernel can boot and exit faster than we can start tailing the logs
+		if opts.Follow && (machine.Status.State == machineapi.MachineStateRunning || machine.Status.State == machineapi.MachineStateExited) {
```
Break this up on multiple lines

craciunoiuc commented:
Hey @andreistan26

any new updates here?

andreistan26 commented:
> Hey @andreistan26
> 
> any new updates here?

Will finish it up this week

craciunoiuc commented:
sure thing, no hurry, just wanted  to get a status update

craciunoiuc commented:
Can you also add this? thanks:

```diff
diff --git a/Makefile b/Makefile
index 2b5d0ac0..18f88386 100644
--- a/Makefile
+++ b/Makefile
@@ -272,6 +272,7 @@ buildenv-myself-full: ## OCI image containing the build environment for KraftKit
 buildenv-myself: ## OCI image containing KraftKit binaries.
 buildenv-qemu: ## OCI image containing a Unikraft-centric build of QEMU.
 buildenv-github-action: ## OCI image used when building Unikraft unikernels in GitHub Actions.
+buildenv-xen: ## OCI image containing a Unikraft-centric build of Xen.
 tools: ## Build all tools.
 kraft: ## The kraft binary.
 runu: ## The runu binary.
```

craciunoiuc commented:
Hey, how is this going

@nderjung should we get to this now and maybe try to get an initial version merged?

@andreistan26 can you do a rebase of this if you're not too busy? thanks

nderjung reviewed (approved):
Generally this is very good, thanks for this awesome contribution!

Please just rebase and we'll get this in as-is.

Cheers

nderjung commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,567 @@
+		return machine, fmt.Errorf("supplied kernel path does not exist: %s", machine.Status.KernelPath)
+	}
+
+	if machine.ObjectMeta.UID == "" {
+		machine.ObjectMeta.UID = uuid.NewUUID()
+	}
```
Following the merge of https://github.com/unikraft/kraftkit/pull/2135

Can you please change this to:
```diff
+ "kraftkit.sh/machine/name"

- 	if machine.ObjectMeta.UID == "" {
- 		machine.ObjectMeta.UID = uuid.NewUUID()
- 	}
+ 	machineID, err := name.NewRandomMachineID()
+ 	if err != nil {
+ 		return fmt.Errorf("could not generate machine UID: %w", err)
+ 	}
```

nderjung commented on machine/xen/v1alpha1.go:
```diff
@@ -0,0 +1,570 @@
+		return machine, fmt.Errorf("supplied kernel path does not exist: %s", machine.Status.KernelPath)
+	}
+
+	machineID, err := name.NewRandomMachineID()
+	if machine.ObjectMeta.UID == "" {
+		return fmt.Errorf("could not generate machine UID: %w", err)
+	}
```
```diff
- 	machineID, err := name.NewRandomMachineID()
- 	if machine.ObjectMeta.UID == "" {
- 		return fmt.Errorf("could not generate machine UID: %w", err)
- 	}
+ 	if machine.ObjectMeta.UID == "" {
+ 		machineID, err := name.NewRandomMachineID()
+ 		if err != nil {
+ 			return fmt.Errorf("could not generate machine UID: %w", err)
+		}
+
+ 		machine.ObjectMeta.UID = machineID
+ 	}
```

nderjung reviewed (changes requested):
Hi, please see this one change.

nderjung reviewed (approved):
Great work, thanks a lot @andreistan26!

olivierlambert commented:
Just wanted to chime in and say congrats, this is great news for the Xen Project :+1:
